### ISSUE TITLE
Buildings with holes & other polygon data

### ISSUE THREAD

**Issue by AtelierLibre** [gboeing/osmnx#223]
Hi Geoff,

I wanted to thank you for OSMnx, it has really started to open up quantitative urban analysis for me.

I'm pretty new to python and coding in general but after following your course information I'm hacking together some osmnx, pandas, geopandas, and shapely to try to quantify urban form (Site Coverage, Floor Area Ratio etc.) using OSM data at the scale of city blocks. I have a couple of issues - one a possible bug, the other a proposal/feature request:

**1. Buildings with holes**
Am I right in thinking that osmnx.buildings currently doesn't handle multi-polygons properly i.e. buildings with holes? In OSM a relation tagged as 'building=yes' with two polygons, one tagged 'role=outer', one tagged 'role=inner' seems to end up in osmnx as two individual building polygons on top of each other. This leads to holes in buildings actually being double counted as building area sometimes generating Site Coverages of more than 1 (see attached image).

**2. Generalise building polygon tools**
I have seen that the osmnx.graph_from_* has been generalised to be able to work with infrastructure other than roads. My question is whether you would consider a similar generalisation for osmnx.buildings to enable it to download and work with other polygonal data for example 'landuse=*' or 'place=*' etc.? I am particularly interested in 'place=city_block' and have currently implemented that for myself by duplicating your building module 'building' with 'city_block' but can see that 'landuse=*' and 'amenity=*' etc. would also be useful.

Kind regards,

Nick Bristow

![coverage_ratio - margery_street](https://user-images.githubusercontent.com/36126654/49305728-d03dd480-f4ae-11e8-8e82-a32b6ab0492f.png)

**Comment by gboeing** [comment]
Regarding 1, that's possible. Would you mind whipping up a MWE code snippet to demonstrate this issue? Regarding 2, that could be a nice idea. Perhaps generalizing it from `buildings` to `footprints`. Pull requests welcome!

**Comment by AtelierLibre** [comment]
Regarding 1. Sure, will do.
Regarding 2, that's great, wanted to check you were open to the idea. In reality it's going to take me a while to generate a pull request with the knowledge I have but would like to give it a go.
All the best,
Nick

**Comment by AtelierLibre** [comment]
# Buildings with holes

I've checked the tagging and this building is tagged as OSM guidelines i.e. building tag on the multipolygon relation which contains two polygons, one tagged role=outer the other tagged role=inner

```
import osmnx as ox
%matplotlib inline

point=(51.52753,-0.10999)

buildings_gdf = ox.buildings_from_point(point=point, distance=25)
bbox = ox.bbox_from_point(point=point, distance=45, project_utm=True)

# project buildings_gdf to UTM for measuring area
buildings_proj = ox.project_gdf(buildings_gdf)
buildings_proj['area']=buildings_proj.area
buildings_proj
```
visualised with

```
fig, ax = ox.plot_buildings(buildings_proj, bgcolor='#333333', color='lightgrey', figsize=(7,7), bbox=bbox,
                            show=True, close=True, dpi=150)
```
shows as  a solid building footprint and 'building_proj' shows two polygons one of 3019.6m2, the other of 1130.6m2 which would give a total of 4150m2

![captura de pantalla de 2018-12-04 17-40-40](https://user-images.githubusercontent.com/36126654/49471513-db18a200-f7eb-11e8-8618-cbed6553b3ce.png)

the same building downloaded from http://overpass-turbo.eu/ as a geojson and projected in QGIS has a footprint of 1889m2 (3019.6m2 - 1130.6m2).

![captura de pantalla de 2018-12-04 17-04-19](https://user-images.githubusercontent.com/36126654/49471351-69405880-f7eb-11e8-9326-724e7ab13e61.png)

Nick

**Comment by gboeing** [comment]
Ah yes. So the buildings module currently doesn't handle relations. It pulls in the data but just treats each as a polygon in the GeoDataFrame that it returns. As far as I can tell, these inner/outer multipolygon relations are relatively rare but they should be handled none the less.

Here (https://github.com/gboeing/osmnx/blob/master/osmnx/buildings.py#L171), we'd need to include a relation check. If there is an inner/outer relation, we'd need to construct the outer polygon, then [difference](https://shapely.readthedocs.io/en/stable/manual.html#object.difference) out each inner polygon. If you'd like to take a crack at this, feel free to open a PR!

**Comment by AtelierLibre** [comment]
Thanks for having a look at it and confirming and also for the pointer in how to fix it. As before, I don't think that I have all the knowledge necessary to do this at the moment.

I could do with a couple of pointers for getting a development environment setup. Am I on the right track if I fork, clone and branch osmnx, make a new conda environment into which I install its dependencies and then run `pip install -e .` to install the local version that I edit?

I may be way out of my depth here so don't worry too much.

Thanks,

Nick

**Comment by gboeing** [comment]
@AtelierLibre yep that development environment setup is correct

**Comment by AtelierLibre** [comment]
@gboeing great thanks

**Comment by gboeing** [comment]
Development happening over in PR #236

**Comment by AtelierLibre** [comment]
@gboeing Regarding point 2 - the generalising of 'buildings' - I have a branch where I have hacked together a working version as a new module called 'footprints' [https://github.com/AtelierLibre/osmnx/tree/Footprints](url)

Can I just check, when you suggested 'footprints' above [https://github.com/gboeing/osmnx/issues/223#issuecomment-443717646](url) was that for a new (largely duplicated) module or just for an new optional argument to the various building functions (similar to 'infrastructure' in the 'graph_from_xxxx' functions)?

I'm guessing the second option would be more concise and easier to maintain but would mean downloading e.g. landuse polygons through a function call similar to `buildings_from_xxxx(footprint='["landuse"]')`

Hope the query is clear, thanks,

Nick

**Comment by gboeing** [comment]
@AtelierLibre it would be to replace the `buildings` module with a new `footprints` module. The `buildings` module should be retained for now, with its functions issuing deprecation warnings when called. The new `footprints` module can duplicate its functionality, but replace "buildings" terminology with "footprints" terminology. The `footprints_from_x` functions could then take a new `footprint_type` argument (with default value of "building") to pass along to the Overpass API here: https://github.com/gboeing/osmnx/blob/master/osmnx/buildings.py#L100

**Comment by AtelierLibre** [comment]
@gboeing great, that's all clear, thanks for the detailed response

**Comment by AtelierLibre** [comment]
Further development in PR [#241](https://github.com/gboeing/osmnx/pull/241)

**Comment by gboeing** [comment]
closed by #236
